modules/reitit-re-frame/src/reitit/re-frame.cljs

```diff
@@ -0,0 +1,111 @@
+(ns reitit.re-frame
+  (:require [re-frame.core :as re :refer [dispatch]]
+            [reitit.frontend :as reitit-frontend]
+            [reitit.core :as reitit]))
+
+(re/reg-event-db ::hash-change
+  (fn [db _]
+    (let [match (reitit-frontend/hash-change (:router (::routes db)) (reitit-frontend/get-hash))]
+      (assoc db ::routes (assoc (::routes db)
+                                :match match
+                                :controllers (if (:enable-controllers? (::routes db))
+                                               (reitit-frontend/apply-controllers (:controllers (::routes db)) match)
+                                               (:controllers (::routes db))))))))
+
+;; Enables controllers (when used the first time)
+;; and applies the changes.
+(re/reg-event-db :routes/apply-controllers
+  (fn [db _]
+    (assoc db ::routes (assoc (::routes db)
+                              :controllers (reitit-frontend/apply-controllers (:controllers (::routes db)) (:match (::routes db)))
+                              :enable-controllers? true))))
+
+;; Options can enable-controllers right away. Should default be enabled or disabled?
+(re/reg-event-fx :routes/init
+  (fn [{:keys [db]} [_ router options]]
+    ;; TODO: This is tied to onhashchange
+    ;; What is user wants to use HTML5 History?
+    (set! js/window.onhashchange #(dispatch [::hash-change]))
+    {:db (assoc db ::routes (merge {:router router
+                                    :match nil
+                                    :controllers []
+                                    :enable-controllers? false}
+                                   options))
+     :dispatch [::hash-change]}))
+
+(re/reg-sub ::state
+  (fn [db]
+    (::routes db)))
+
+(re/reg-sub :routes/match
+  :<- [::state]
+  (fn [state]
+    (:match state)))
+
+(re/reg-sub :routes/data
+  :<- [:routes/match]
+  (fn [match _]
+    (:data match)))
+
+(re/reg-sub :routes/router
+  :<- [::state]
+  (fn [state]
+    (:router state)))
+
+(re/reg-sub :routes/match-by-name
+  :<- [:routes/router]
+  (fn [router [_ k params]]
+    (if router
+      (reitit/match-by-name router k params)
+      ::not-initialized)))
+
+(defn routed-view
+  "Renders the current view component.
+
+  Current view component is the :view from last :controllers entry with :view set."
+  []
+  (let [view @(re/subscribe [:routes/match])
+        ;; Select downmost controller that has view component
+        controller (first (filter :view (reverse (:controllers (:data view)))))]
```

Comment:
Today I think the `:view` should not be part of the Controller, it should be normal route data instead. With the current solution, the last view masks all previous ones. If there are multiple controllers in an endpoint, the order matters. e.g.

```clj
["/admin/users" {:controllers [users-controller admin-controller]}]
```

```clj
["/admin/users" {:controllers [admin-controller users-controller]}]
```

might render totally different uis. Instead:

```clj
["/admin/users" {:view users-view
                 :controllers [users-controller admin-controller]}]
```